README
This is PARI/GP, version 2.12.x (*).

PARI/GP Number Theory-oriented Computer Algebra System
        Copyright (C) 2000-2019 The PARI Group, Bordeaux.

==========================================================================

To get started, run

  cd doc; tex INSTALL.tex

and have a look at the file INSTALL.dvi, extracted verbatim from Appendix A
of the User Manual. (Or pdftex INSTALL.tex if PDF files are more convenient.)
If you are in a hurry, look at the ASCII file 'INSTALL' but a lot of
information is missing there.

Once 'Configure' has been run, 'make doc' typesets separate user's manual for
GP and libpari, a tutorial and reference card in directory doc.

The file NEW contains new features with respect to previous stable versions.
The file COMPAT warns about incompatible changes. See the files
CHANGES* for a terse description of successive patches.

For the adventurous: to check out bleeding-edge development versions of
PARI/GP, see the file README-git.

Afterwards, you're on your own.

==========================================================================

There are three mailing lists devoted to the PARI/GP package, and most
feedback should be directed to those. They are:

* pari-announce: to announce major version changes. You can't write to this
one, but you should probably subscribe.

* pari-dev: for everything related to the development of PARI, including
suggestions, technical questions, bug reports or patch submissions.

* pari-users: for everything else.

To subscribe, send empty messages respectively to

   pari-announce-request@pari.math.u-bordeaux.fr
   pari-users-request@pari.math.u-bordeaux.fr
   pari-dev-request@pari.math.u-bordeaux.fr

with a Subject: containing the word "subscribe".

If you are not a member of any of those lists and do not want to become
one, you can write to us at

  pari@math.u-bordeaux.fr

We may forward your mail to the lists above and will definitely try to
correct faulty behaviour, if necessary. But we cannot promise that you
will get an individual answer.

Last but not least, PARI home page can be found at

  http://pari.math.u-bordeaux.fr/

  Thanks for your support, and have fun !

========================================================================

PARI/GP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


(*) Version numbers have the shape: MAJOR.MINOR.patchlevel.status. The status
increases through {alpha, beta, release}. New features are introduced during
the `alpha' phase, tested during `beta', and declared stable when `release'
is reached.
The MAJOR version number will not change for a while.
The MINOR version number increases by 2 when a new status cycle starts (as
soon as a 'release' version is out): odd numbers are reserved for 'released'
versions, even ones for alpha/beta phase.
The patchlevel goes up each time we feel an upgrade is necessary (improvement
or bugfix). 'Release' versions may be updated with further patchlevels if
important bugs need to be fixed before the next cycle reaches completion.


CHANGES
Bug numbers refer to the BTS at http://pari.math.u-bordeaux.fr/Bugs/
Done for version 2.14.0 (released ??/??/2020):
  Fixed
     1- slowdown in zeta (e.g. zeta(Pi) at \p500)
     2- GC error in idealinv [#2258]
     3- qfminim inefficiency (initial bound from non-LLL-reduced matrix)
     4- mfshimura in weight 3/2 => oo loop
     5- zetahurwitz(s, huge x) => overflow or very slow
     6- matsolve([1,1,1/2; 1,0,1/2; 2,1,0], [-1;1;-1]) -> SEGV [#2261]
     7- qfminim(Mat(1),,,2) -> precision error [#2262]
     8- subst(p, v, t_SER) very slow if p contains many variables [#2264]
     9- mfsymboleval for trivial path returned 0 instead of a vector of 0s
        when f has multiple embeddings [#2260]
    10- 2^(1 + O(3^4)) -> type error
    11- Zn_quad_roots(8,0,-1) to compute roots of x^2-1 mod 8 -> [4,[1,3]]
        (1 or 3 mod 4) instead of the expected [2,[1]] (1 mod 2) [#2265]
    12- tan(1+10^20*I) -> overflow (same for cotan and psi)
BA  13- Mod(2,3)^1000000000000000000001 -> Mod(-1,3) [#2266]

  Added
HC   1- new GP function eulerreal
BA   2- fplll: fast, dpe and heuristic variants
     3- dirpowerssum: allow a completely multiplicative function
     4- [libpari] atanhuu, atanhui, powcx, powcx_prec
     5- new GP function lfunparams (initial implementation P. Molin)
     6- [libpari] row_Q_primpart

  Changed
HC   1- faster implementation of eulervec + cache result
     2- sumnummonieninit(,1): faster algorithm (complex step)
     3- atanh(rational) now uses binary splitting
     4- prototype of FqV_to_FlxV, FqC_to_FlxC, FqM_to_FlxM: argument 'p' is
        an ulong, no longer a GEN

Done for version 2.13.0 (released 19/10/2020):
  Fixed
     1- sizebyte(objects containing t_LIST): would assign size 3 to all lists
     2- [Map(Mat([0, 0])), [0, 0, 1]] -> SEGV
     3- zeta(I+2^(-64)) -> SEGV
BA   4- permorder returned incorrect results for large orders
     5- p=[4,54]; lfunmf(mfinit(p,0), mftraceform(p,0)) -> oo loop
BA   6- printf("%5.2f",M) transposed M [#2252]
     7- bnfinit instability at low accuracy. E.g., at \p38
        P=x^7+22*x^6+184*x^5+764*x^4+2020*x^3+3792*x^2+5712*x+4;
        bnfinit(P,1).no -> 4 times too large
     8- accuracy problems in bnfisunit after bnfinit(P,1). E.g.,
        P=x^7-22*x^6+169*x^5-568*x^4+1263*x^3-1506*x^2+2295*x+2;
        setrand(1); K=bnfinit(P,1); U=bnfunits(K,idealprimedec(K,2));
        bnfisunit(K,4*(x^2-18*x+85)/x^2*Mod(1,K.pol),U) -> error

  Added
     1- examples/pari.f90: example FORTRAN binding
     2- [libpari] quodif
BA   3- [libpari] perm_sqr, perm_powu, perm_orderu
BA   4- asymptotically fast quadunit

  Changed
     1- faster intnumgaussinit + change default number of points to B/4,
        suitable when b - a ~ distance to the nearest pole
BA   2- [libpari] perm_pow now take a t_INT exponent
BA   3- [libpari] permorder and perm_order now return a t_INT.

Done for version 2.12.1 (released 16/09/2020):
  Fixed
BA   1- plotlines flag was ignored
     2- fileopen(,"w" or "a") was bypassing the check for 'secure' default
     3- bug in idealismaximal: K=nfinit(y^2+7); idealismaximal(K,[11,7;0,1])->0
     4- bnfcertify did not certify correctly its argument when DEBUGLEVEL > 1
        (\g 2 and above) and the computed class group was trivial (introduced
        in 2002, commit d3d44841). The code was correct at \g1 or \g0, and in
        all cases if the computed class group was non-trivial.
     5- nfdiscfactors(nfinit([x^2+1, [1,x]])) -> wrong result (didn't support
        inputs from nfinit([T, integer basis]))
     6- nfinit(rnf) would produce a broken nf (not supporting
        nf_get_ramified_primes)
     7- [mf,F] = mffromqf([2,0;0,2]); L=lfunmf(mf,F) => wrong values
     8- t=mfTheta();mfslashexpansion(mfinit(t),t,[0,1;-1,0],1,1) -> error
     9- poldisc(x^4 + 8/y*x) -> type error in Q_divi_to_int [#2157]
    10- T=mfTheta(); mfeval(mfinit(T),T,I/10) -> wrong sign
    11- mfslashexpansion(mf,F,huge matrix,...) -> overflow
    12- mf=mfinit([470,2],0);F=mfeigenbasis(mf)[7]; mfcuspval(mf,F,1/5)
        -> division by 0
    13- quadtofp(quadgen(-3)) -> t_FRAC component [#2158]
    14- [from 2.12.0] nfgcd wouldn't terminate [#2159]
    15- conjvec(t_FFELT of characteristic 2) was missing a conjugate
    16- nf2 = nfnewprec(nf) could return incompatible nf2 [#2154]
    17- [from 2.12.0] u_forprime_next around 2^63 could return primes larger
        than upper bound
    18- [from 2.12.0] gcd(2*x, 2*x+1) -> 2 [instead of 1]
    19- [from 2.12.0] deriv(2/(4*x^2+1),x)->inconsistent exact division [#2162]
    20- issquare(non square t_FFELT, &x) -> SEGV [#2163]
    21- rdivii / rdiviiz imprecise, e.g.
        a=28088408347805994933529600180306266460976;
        b=10333147966386144929666651337523200000000; rdivii(a, b)
    22- typo in QM_inv (wrong result)
    23- modular symbols package: better type checks
    24- division by 0 in Flx_extresultant when one input is 0
    25- [from 2.12.0] lfunconductor(1,[]) -> SEGV
    26- internal error in isprime(,1) after addprime() [#2166]
    27- ZM_gauss returned incorrect result for t_MAT RHS and 0 input
BA  28- lfunhardy(..., t) with t<0 -> error
BA  29- ellweilpairing(E,P,P, odd m) -> wrong result [#2172]
PB  30- FlxqM_mul could return wrong result [#2174]
    31- p = 4*vecprod(primes([3,373]))*587-1; quadclassunit(-p) -> SEGV [#2175]
BA  32- Using local in export()ed functions could crash GP
    33- Flx_div_by_X_x(A, x) didn't handle deg(A) <= 0
    34- M=mfinit([11,2,Mod(3,11)]); f=mfbasis(M)[1]; lfuncheckfeq(lfunmf(M,f))
        -> type error
    35- rare random SEGV in nfbasis [#2178]
    36- missing type checks in lfunthetacheckinit [#2179]
    37- idealispower(x,n,&y): y could be incorrect [#2181]
    38- SEGV in idealispower [#2183]
    39- inconsistencies in subst for constant polynomials: e.g.,
          type(subst('u^0,u,x)) -> t_POL (was t_INT)
          subst('u^0,u,Mat(Mod(1,2))) -> [Mod(1,2)] (was [1])

src/gp/gp.c
/* Copyright (C) 2000  The PARI group.

This file is part of the PARI/GP package.

PARI/GP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA. */

/*******************************************************************/
/**                                                               **/
/**                        PARI CALCULATOR                        **/
/**                                                               **/
/*******************************************************************/
#ifdef _WIN32
#  include "../systems/mingw/pwinver.h"
#  include <windows.h>
#  include "../systems/mingw/mingw.h"
#endif
#include "pari.h"
#include "paripriv.h"
#include "gp.h"

static jmp_buf *env;
static pari_stack s_env;
void (*cb_gp_output)(GEN z) = NULL;
void (*cb_pari_end_output)(void) = NULL;

static void
gp_ask_confirm(const char *s)
{
  err_printf(s);
  err_printf(". OK ? (^C if not)\n");
  pari_hit_return();
}

/* numerr < 0: after changing PARI stack size
 * numerr > 0: normal error, including SIGINT */
static void
gp_err_recover(long numerr)
{
  longjmp(env[s_env.n-1], numerr);
}

/* numerr >= 0 */
static void
gp_pre_recover(long numerr)
{
  out_puts(pariErr, "\n"); pariErr->flush();
  gp_err_recover(numerr);
}

static void
reset_ctrlc(void)
{
#if defined(_WIN32) || defined(__CYGWIN32__)
  win32ctrlc = 0;
#endif
}

static int
is_silent(char *s) { return s[strlen(s) - 1] == ';'; }

static int stdin_isatty = 0;
static int
is_interactive(void)
{ return pari_infile == stdin && stdin_isatty; }

/*******************************************************************/
/**                                                               **/
/**                        INITIALIZATION                         **/
/**                                                               **/
/*******************************************************************/
static void
print_shortversion(void)
{
  const ulong mask = (1UL<<PARI_VERSION_SHIFT) - 1;
  ulong n = paricfg_version_code, major, minor, patch;

  patch = n & mask; n >>= PARI_VERSION_SHIFT;
  minor = n & mask; n >>= PARI_VERSION_SHIFT;
  major = n;
  printf("%lu.%lu.%lu\n", major,minor,patch); exit(0);
}

static void
usage(char *s)
{
  printf("### Usage: %s [options] [GP files]\n", s);
  printf("Available options:\n");
  printf("  [-f,--fast]\t\tFast start: do not read .gprc\n");
  printf("  [-q,--quiet]\t\tQuiet mode: do not print banner and history numbers\n");
  printf("  [-s stacksize]\tStart with the PARI stack of given size (in bytes)\n");
  printf("  [--default key=val]\tExecute default(key,val) on startup\n");
  printf("  [--emacs]\t\tRun as if in Emacs shell\n");
  printf("  [--help]\t\tPrint this message\n");
  printf("  [--test]\t\tTest mode. No history, wrap long lines (bench only)\n");
  printf("  [--texmacs]\t\tRun as if using TeXmacs frontend\n");
  printf("  [--version]\t\tOutput version info and exit\n");
  printf("  [--version-short]\tOutput version number and exit\n\n");
  exit(0);
}

static void
gp_head(void)
{
  pari_print_version();
  pari_putc('\n');
  pari_center("Copyright (C) 2000-2020 The PARI Group");
  pari_putc('\n');
  print_text("PARI/GP is free software, covered by the GNU General Public \
License, and comes WITHOUT ANY WARRANTY WHATSOEVER.");
  pari_puts("\nType ? for help, \\q to quit.\n");
  pari_printf("Type ?%d for how to get moral"
              " (and possibly technical) support.\n", pari_community());
  if (pari_mainstack->vsize)
    pari_printf("\nparisizemax = %lu, primelimit = %lu",
                pari_mainstack->vsize,GP_DATA->primelimit);
  else
    pari_printf("\nparisize = %lu, primelimit = %lu",
                pari_mainstack->rsize,GP_DATA->primelimit);
  if (pari_mt_nbthreads > 1)
    pari_printf(", nbthreads = %lu", pari_mt_nbthreads);
  pari_putc('\n');
}

static char *
read_arg(long *nread, char *t, long argc, char **argv)
{
  long i = *nread;
  if (isdigit((int)*t)) return t;
  if (*t || i==argc) usage(argv[0]);
  *nread = i+1; return argv[i];
}

static char *
read_arg_equal(long *nread, char *t, long argc, char **argv)
{
  long i = *nread;
  if (*t=='=' && isdigit((int)t[1])) return t+1;
  if (*t || i==argc) usage(argv[0]);
  *nread = i+1; return argv[i];
}

static void
init_trivial_stack(void)
{
  const size_t s = 2048;
  pari_mainstack->size = s;
  pari_mainstack->bot = (pari_sp)pari_malloc(s);
  avma = pari_mainstack->top = pari_mainstack->bot + s;
}

static void
free_trivial_stack(void)
{
  free((void*)pari_mainstack->bot);
}

typedef struct { char *key, *val; } pair_t;
/* If ab of the form key=val, record pair in new stack entry
 * P[n].key must be freed by caller to avoid memory leak */
static void
record_default(pari_stack *s_P, char *ab)
{
  pair_t *P = (pair_t*)*pari_stack_base(s_P);
  char *k, *v;
  long n;
  ab = pari_strdup(ab);
  parse_key_val(ab, &k, &v);
  n = pari_stack_new(s_P);
  P[n].key = k;
  P[n].val = v;
}
static void
read_opt(pari_stack *p_A, long argc, char **argv)
{
  pair_t *P;
  pari_stack s_P; /* key / value to record default() settings */
  char *b = NULL, *p = NULL, *s = NULL;
  ulong f = GP_DATA->flags;
  long i = 1, initrc = 1;

  (void)&p; (void)&b; (void)&s; /* -Wall gcc-2.95 */

  pari_stack_init(&s_P,sizeof(*P),(void**)&P);
  pari_stack_alloc(&s_P, 64);
  pari_outfile = stderr;
  while (i < argc)
  {
    char *t = argv[i];

    if (*t++ != '-') break;
    i++;
START:
    switch(*t++)
    {
      case 'p': p = read_arg(&i,t,argc,argv); break;
      case 's': s = read_arg(&i,t,argc,argv); break;
      case 'e':
        f |= gpd_EMACS; if (*t) goto START;
        break;
      case 'q':
        f |= gpd_QUIET; if (*t) goto START;
        break;
      case 't':
        f |= gpd_TEST; if (*t) goto START;
        break;
      case 'f':
        initrc = 0; if (*t) goto START;
        break;
      case 'D':
        if (*t || i == argc) usage(argv[0]);
        record_default(&s_P, argv[i++]);
        break;
      case '-':
        if (strcmp(t, "version-short") == 0) { print_shortversion(); exit(0); }
        if (strcmp(t, "version") == 0) {
          init_trivial_stack(); pari_print_version();
          free_trivial_stack(); exit(0);
        }
        if (strcmp(t, "default") == 0) {
          if (i == argc) usage(argv[0]);
          record_default(&s_P, argv[i++]);
          break;
        }
        if (strcmp(t, "texmacs") == 0) { f |= gpd_TEXMACS; break; }
        if (strcmp(t, "emacs") == 0) { f |= gpd_EMACS; break; }
        if (strcmp(t, "test") == 0) { f |= gpd_TEST; initrc = 0; break; }
        if (strcmp(t, "quiet") == 0) { f |= gpd_QUIET; break; }
        if (strcmp(t, "fast") == 0) { initrc = 0; break; }
        if (strncmp(t, "primelimit",10) == 0) {p = read_arg_equal(&i,t+10,argc,argv); break; }
        if (strncmp(t, "stacksize",9) == 0) {s = read_arg_equal(&i,t+9,argc,argv); break; }
       /* fall through */
      default:
        usage(argv[0]);
    }
  }
  if (f & gpd_TEST) stdin_isatty = 0;
  GP_DATA->flags = f;
#ifdef READLINE
  GP_DATA->use_readline = stdin_isatty;
#endif
  if (!is_interactive()) GP_DATA->breakloop = 0;
  if (initrc) gp_initrc(p_A);
  for ( ; i < argc; i++) pari_stack_pushp(p_A, pari_strdup(argv[i]));

  /* override the values from gprc */
  if (p) (void)sd_primelimit(p, d_INITRC);
  if (s) (void)sd_parisize(s, d_INITRC);
  for (i = 0; i < s_P.n; i++) {
    setdefault(P[i].key, P[i].val, d_INITRC);
    free((void*)P[i].key);
  }
  pari_stack_delete(&s_P);
  pari_outfile = stdout;
}

/*******************************************************************/
/**                                                               **/
/**                            TEST MODE                          **/
/**                                                               **/
/*******************************************************************/
static int
test_is_interactive(void) { return 0; }

static void
test_output(GEN z) { init_linewrap(76); gen_output(z); }
void
init_test(void)
{
  disable_color = 1;
  init_linewrap(76);
  pari_errfile = stdout;
  cb_gp_output = test_output;
  cb_pari_is_interactive = test_is_interactive;
}

/*******************************************************************/
/**                                                               **/
/**                   FORMAT GP OUTPUT                            **/
/**                                                               **/
/*******************************************************************/
    /* REGULAR */
static void
normal_output(GEN z, long n)
{
  long l = 0;
  char *s;
  /* history number */
  if (n)
  {
    char buf[64];
    if (!(GP_DATA->flags & gpd_QUIET))
    {
      term_color(c_HIST);
      sprintf(buf, "%%%ld = ", n);
      pari_puts(buf);
      l = strlen(buf);
    }
  }
  /* output */
  term_color(c_OUTPUT);
  s = GENtostr(z);
  if (GP_DATA->lim_lines)
    lim_lines_output(s, l, GP_DATA->lim_lines);
  else
    pari_puts(s);
  pari_free(s);
  term_color(c_NONE); pari_putc('\n');
}

static void
gp_output(GEN z)
{
  if (cb_gp_output) { cb_gp_output(z); return; }
  if (GP_DATA->fmt->prettyp == f_PRETTY)
  { if (tex2mail_output(z, GP_DATA->hist->total)) return; }
  normal_output(z, GP_DATA->hist->total);
  pari_flush();
}

static GEN
gp_main_loop(long ismain)
{
  VOLATILE GEN z = gnil;
  VOLATILE long t = 0, r = 0;
  VOLATILE pari_sp av = avma;
  filtre_t F;
  Buffer *b = filtered_buffer(&F);
  struct gp_context rec;
  long er;
  if ((er = setjmp(env[s_env.n-1])))
  { /* recover: jump from error [ > 0 ] or allocatemem [ -1 ] */
    if (er > 0) { /* true error */
      if (!(GP_DATA->recover)) exit(1);
      gp_context_restore(&rec);
      /* true error not from main instance, let caller sort it out */
      if (!ismain) { kill_buffers_upto_including(b); return NULL; }
    } else { /* allocatemem */
      tmp_restore(rec.file.file);
      gp_context_save(&rec);
    }
    set_avma(av = pari_mainstack->top);
    parivstack_reset();
    kill_buffers_upto(b);
    pari_alarm(0);
  }
  for(;;)
  {
    gp_context_save(&rec);
    if (! gp_read_line(&F, NULL))
    {
      if (popinfile()) gp_quit(0);
      if (ismain) continue;
      pop_buffer(); return z;
    }
    if (ismain)
    {
      reset_ctrlc();
      timer_start(GP_DATA->T);
      walltimer_start(GP_DATA->Tw);
      pari_set_last_newline(1);
    }
    if (gp_meta(b->buf,ismain)) continue;
    z = pari_compile_str(b->buf);
    z = closure_evalres(z);
    if (!ismain) continue;

    t = timer_delay(GP_DATA->T);
    r = walltimer_delay(GP_DATA->Tw);
    if (!pari_last_was_newline()) pari_putc('\n');
    pari_alarm(0);
    if (t && GP_DATA->chrono)
    {
      if (pari_mt_nbthreads==1)
      {
        pari_puts("time = ");
        pari_puts(gp_format_time(t));
      }
      else
      {
        pari_puts("cpu time = ");
        pari_puts(gp_format_time(t));
        pari_puts(", real time = ");
        pari_puts(gp_format_time(r));
      }
      pari_puts(".\n");
    }
    if (GP_DATA->simplify) z = simplify_shallow(z);
    pari_add_hist(z, t, r);
    if (z != gnil && ! is_silent(b->buf) ) gp_output(z);
    set_avma(av);
    parivstack_reset();
  }
}

/* as gp_read_file, before running the main gp instance */
static void
read_main(const char *s)
{
  GEN z;
  if (setjmp(env[s_env.n-1]))
    z = NULL;
  else {
    FILE *f = switchin(s);
    if (file_is_binary(f)) {
      z = readbin(s,f, NULL);
      popinfile();
    }
    else z = gp_main_loop(0);
  }
  if (!z) err_printf("... skipping file '%s'\n", s);
  set_avma(pari_mainstack->top);
}

static const char *
break_loop_prompt(long n)
{
  const char *s = (n==1)? "break> ": stack_sprintf("break[%ld]> ", n);
  return gp_format_prompt(s);
}

static long frame_level=0, dbg_level = 0;

static int
break_loop(int numerr)
{
  filtre_t F;
  Buffer *b;
  int sigint = numerr<0, go_on = sigint;
  struct gp_context rec1, rec2;
  const char *prompt, *msg;
  long nenv, oldframe_level = frame_level;
  pari_sp av;

  if (numerr == e_SYNTAX) return 0;
  if (numerr == e_STACK) { evalstate_clone(); set_avma(pari_mainstack->top); }
  gp_context_save(&rec1);

  b = filtered_buffer(&F);
  nenv=pari_stack_new(&s_env);
  prompt = break_loop_prompt(s_env.n-1);
  iferr_env = NULL;
  dbg_level = 0;
  frame_level = closure_context(oldframe_level, dbg_level);
  pari_infile = newfile(stdin, "stdin", mf_IN)->file;
  term_color(c_ERR); pari_putc('\n');
  if (sigint)
    msg = "Break loop: <Return> to continue; 'break' to go back to GP prompt";
  else
    msg = "Break loop: type 'break' to go back to GP prompt";
  print_errcontext(pariOut, msg, NULL, NULL);
  term_color(c_NONE);
  av = avma;
  for(;;)
  {
    GEN x;
    long er, br_status;
    set_avma(av);
    gp_context_save(&rec2);
    if ((er=setjmp(env[nenv])))
    {
      if (er < 0)
      {
        s_env.n = 1;
        frame_level = oldframe_level;
        longjmp(env[s_env.n-1], er);
      }
      gp_context_restore(&rec2);
      iferr_env = NULL;
      closure_err(dbg_level);
      compilestate_restore(&rec1.eval.comp);
      (void) closure_context(oldframe_level, dbg_level);
      pari_infile = newfile(stdin, "stdin", mf_IN)->file;
    }
    term_color(c_NONE);
    if (!gp_read_line(&F, prompt))
      br_status = br_BREAK; /* EOF */
    else
    {
      /* Empty input ? Continue if entry on sigint (exit debugger frame) */
      if (! *(b->buf) && sigint) break;
      reset_ctrlc();
      if (gp_meta(b->buf,0)) continue;
      x = pari_compile_str(b->buf);
      x = closure_evalbrk(x, &br_status);
    }
    switch (br_status)
    {
      case br_NEXT: case br_MULTINEXT:
        popinfile(); /* exit frame. Don't exit debugger if s_env.n > 2 */
        go_on = 0; goto BR_EXIT;
      case br_BREAK: case br_RETURN:
        killallfiles(); /* completely exit the debugger */
        go_on = 0; goto BR_EXIT;
    }
    if (x!=gnil && !is_silent(b->buf)) { term_color(c_OUTPUT); gen_output(x); }
  }
BR_EXIT:
  s_env.n=nenv;
  frame_level = oldframe_level;
  gp_context_restore(&rec1);
  pop_buffer(); return go_on;
}

#ifdef __CYGWIN32__
void
cyg_environment(int argc, char ** argv)
{
  char *ti_dirs = getenv("TERMINFO_DIRS");
  char *argv0, *p;
  char *newdir;
  long n;

  if (!argc || !argv) return;
  argv0 = *argv;
  if (!argv0 || !*argv0) return;
  p = strrchr(argv0, '/');
  if (!p)
    p = argv0 = "";
  else
    p++;
  n = p - argv0;
  if (ti_dirs)
  {
    n += 14 + strlen(ti_dirs) + 1 + 8 + 1;
    newdir = malloc(n);
    if (!newdir) return;
    snprintf(newdir, n-8, "TERMINFO_DIRS=%s:%s", ti_dirs, argv0);
  }
  else
  {
    n += 14 + 8 + 1;
    newdir = malloc(n);
    if (!newdir) return;
    snprintf(newdir, n-8, "TERMINFO_DIRS=%s", argv0);
  }
  strcpy(newdir+n-9,"terminfo");
  putenv(newdir);
}
#endif

int
main(int argc, char **argv)
{
  char **A;
  pari_stack s_A;

  GP_DATA = default_gp_data();
  pari_stack_init(&s_env, sizeof(*env), (void**)&env);
  (void)pari_stack_new(&s_env);

  if (setjmp(env[s_env.n-1]))
  {
    puts("### Errors on startup, exiting...\n\n");
    exit(1);
  }
#ifdef __CYGWIN32__
  cyg_environment(argc, argv);
#endif
  stdin_isatty = pari_stdin_isatty();
  pari_init_defaults();
  pari_library_path = DL_DFLT_NAME;
  pari_stack_init(&s_A,sizeof(*A),(void**)&A);
  /* must be defined here in case an error is raised in pari_init_opts, e.g.
   * when parsing function prototypes */
  cb_pari_err_recover = gp_err_recover;
  pari_init_opts(1000000 * sizeof(long), 0, INIT_SIGm | INIT_noPRIMEm | INIT_noIMTm);
  cb_pari_pre_recover = gp_pre_recover;
  cb_pari_break_loop = break_loop;
  cb_pari_is_interactive = is_interactive;

  read_opt(&s_A, argc,argv);
  pari_init_primes(GP_DATA->primelimit);
#ifdef SIGALRM
  (void)os_signal(SIGALRM,gp_alarm_handler);
#endif
  pari_add_module(functions_gp);

  pari_set_plot_engine(gp_get_plot);
  cb_pari_quit = gp_quit;
  cb_pari_whatnow = whatnow;
  cb_pari_sigint = gp_sigint_fun;
  cb_pari_handle_exception = gp_handle_exception;
  cb_pari_ask_confirm = gp_ask_confirm;
  pari_init_paths();
  pari_mt_init(); /* MPI: will not return on slaves (pari_MPI_rank = 0) */
#ifdef _WIN32
  if (stdin_isatty) win32_set_codepage();
#endif
#ifdef READLINE
  init_readline();
#endif
  if (GP_DATA->flags & gpd_EMACS) init_emacs();
  if (GP_DATA->flags & gpd_TEXMACS) init_texmacs();

  timer_start(GP_DATA->T);
  walltimer_start(GP_DATA->Tw);
  if (!(GP_DATA->flags & gpd_QUIET)) gp_head();
  if (GP_DATA->flags & gpd_TEST) init_test();
  if (s_A.n)
  {
    FILE *l = pari_logfile;
    long i;
    pari_logfile = NULL;
    for (i = 0; i < s_A.n; pari_free(A[i]),i++) read_main(A[i]);
    /* Reading one of the input files above can set pari_logfile.
     * Don't restore in that case. */
    if (!pari_logfile) pari_logfile = l;
  }
  pari_stack_delete(&s_A);
  (void)gp_main_loop(1);
  gp_quit(0);
  return 0; /* LCOV_EXCL_LINE */
}

void
pari_breakpoint(void)
{
  if (!pari_last_was_newline()) pari_putc('\n');
  closure_err(0);
  if (cb_pari_break_loop && cb_pari_break_loop(-1)) return;
  cb_pari_err_recover(e_MISC);
}

void
dbg_down(long k)
{
  if (k<0) k=0;
  dbg_level -= k;
  if (dbg_level<0) dbg_level=0;
  gp_err_recover(e_NONE);
}

GEN
dbg_err(void) { GEN E = pari_err_last(); return E? gcopy(E):gnil; }

void
dbg_up(long k)
{
  if (k<0) k=0;
  dbg_level += k;
  if (dbg_level>frame_level) dbg_level=frame_level;
  gp_err_recover(e_NONE);
}

void
gp_quit(long code)
{
  pari_kill_plot_engine();
  pari_close();
  kill_buffers_upto(NULL);
  if (!(GP_DATA->flags & gpd_QUIET)) pari_puts("Goodbye!\n");
  if (cb_pari_end_output) cb_pari_end_output();
  exit(code);
}

void
whatnow0(char *s) { whatnow(pariOut, s,0); }

#include "gp_init.h"
